ctapipe/io/tableloader.py

```diff
@@ -119,9 +134,173 @@ def _read_telescope_table(self, group, tel_id):
+            but only event_id-based quantities have been selected together with it.""")
+            self.log.critical("No data enabled for loading.")
+            self.close()
+            return None # it should probably return some exit status != 0
+        
+        split_mode = get_structure(self.h5file)
+        
+        if labels is not None:
+            if all([isinstance(label, str) for label in labels]):
+                selected_tel_types = labels
+                selected_tel_ids = []
+                for tel_type in labels:
+                    selected_tel_ids += self.subarray.get_tel_ids_for_type(tel_type)
+            elif all([isinstance(label, int) for label in labels]):
+                selected_tel_ids = labels
+            else:
+                self.log.error("Check supplied labels to 'read_data'")
+        
+        if split_mode == "by_id":
+
+            output = self.read_telescope_events(tel_ids=selected_tel_ids)
+            
+            try: 
+                output.sort(sort_by)
+            except ValueError:
+                output.sort(["tel_id"])
+        
+        elif split_mode == "by_type":
+            
+            output = self.read_telescope_events_by_tel_type(tel_ids = selected_tel_ids)
+        
+            try:
+                for tel_type in output.keys():
+                    output[tel_type].sort(sort_by)
+            except ValueError:
+                for tel_type in output.keys():
+                    output[tel_type].sort(["tel_id"])
+        
+        self.close()
+        
+        return output
+    
+    def read_telescope_events_by_tel_type(self, tel_ids=None):
+                
+        selected_subarray = self.subarray.select_subarray(tel_ids, name="")
+        selected_tel_types = selected_subarray.telescope_types
+        
+        table_dict = {}
+        for tel_type in selected_tel_types:
+            
+            selected_tel_ids = [tel_id for tel_id in tel_ids if tel_id in self.subarray.get_tel_ids_for_type(tel_type)]
+            table_dict[str(tel_type)] = self.read_telescope_events(selected_tel_ids)
+        
+        return table_dict
+
+    def read_telescope_events(self, tel_ids=None):
+        if tel_ids is None:
+            tel_ids = self.subarray.tel.keys()
+        
         table = None
+        
+        # Joini all the DL1-based data, which depends on the triggered tel_ids
+        if (self.load_dl1_images or self.load_dl1_parameters or self.load_true_images) and ((table is None) or (len(table)==0)):
+            table = vstack([self.read_telescope_events_for_id(tel_id) for tel_id in tel_ids])
+        elif (self.load_dl1_images or self.load_dl1_parameters or self.load_true_images) and (len(table) > 0):
+            table = join(
+                table,
+                vstack([self.read_telescope_events_for_id(tel_id, table) for tel_id in tel_ids]),
+                keys=["tel_id"],
+                join_type="inner",
```

Comment:
Are you sure you always want *inner* joins? That would skip rows where one or the other is missing, rather than keep the existing one and give Null values for the other.  It may be ok, but I wonder if there might be cases where we would silently drop events with this.

Reply:
need to check thanks, for now it seems to work...

Reply:
perhaps you can check out this PR and test it with `ctapipe-process` in case I messed something